Match IO handlers' port size to its actual size

kcgen opened the pull request:
This PR reduces the IO hander type's **port** size from `Bitu` to 16-bits.  

IO handers are function pointers that are stored in six arrays (three IO-writer arrays of 64k elements, and three IO-reader arrays of 64k elements), queried every 1ms of emulation time for input and/or output. The port value is an offset into these handler arrays.

In real-world use cases, typically there are on the order of ~100s IO handers registered at any time (Tandy writer, Sound blaster writer, CD-DA callback writer, serial ports reading and writing, and so on). The port value does not exceed a couple thousand (and logically, if it ever did exceed 16-bits it would overflow the handler arrays).

Besides reducing the carrying weight by 6-bytes across 6 * 64k elements, this allows compilers to more readily bundle the port into SIMD operations.  For example, NEON instructions allow packing 4 x 16-bit values together; where as the port's prior 64-bit weight excluded it from NEON all together.

No formatting has been applied. 

This change can be programmatically applied to the code using:

- Definition change:

  ``` shell
  git grep -l 'Bitu [/\*]*port[\*/]*' | xargs sed -i 's|Bitu [/\*]*port[/\*]*|uint16_t port|g'
  sed -i 's/uint16 port=decode_fetchb/Bit8u port=decode_fetchb/g' src/cpu/core_dyn_x86/decoder.h  # revert four changes accidentally having similar pattern but not related
  ```

- Argument-size change in DMA, AdLib, Sound Blaster, and Serial Port classes:

  ``` shell
  cd src/hardware
  sed -i 's/Bitu i;/uint16_t i;/g' dma.cpp
  sed -i 's/Bitu base/uint16_t base/g' adlib.cpp sblaster.cpp 
  sed -i 's/for (uint32_t i = 0; i <= 7;/for (uint16_t i = 0; i <= 7;/' serialport/serialport.cpp 
  ```

- Printf-format change (also inside `src/hardware`):

  ``` shell
  sed -i 's/port %#" PRIxPTR/port %#" PRIx16/g; s/xPTR, port/x16, port/g' iohandler.cpp tandy_sound.cpp vga_xga.cpp
  ```

dreamer commented:
Before this draft will turn into a PR with many hours sunk into it: please don't pursue this.

1. In `inout.h` IO_MAX is defined as (64*1024+3) > 2<sup>16</sup>
2. In several places I noticed 16-bit port is then packaged back into Bitu for processing by other functions
3. I don't see think this compiler will manage to optimize this code for SIMD - I see no parallel provessing in here
4. This change is quite invasive and will be hard to review, especially since I don't know IO code yet.
5. I can't rule out, that using Bitu in this case might be a valid micro-optimization (I don't know if it is, I just can't rule it out)

kcgen commented:
> In `inout.h` IO_MAX is defined as (64*1024+3) > 2^16

Yes, the six arrays are each +3 elements longer than 2^16, but those three are manually added from the port value (so `port` itself can still be 16-bit).  `IO_MAX` is left as-is.

src/hardware/iohandler.cpp:   io_writehandlers[0][port+0](port+0,(val >> 0) & 0xff,1);
src/hardware/iohandler.cpp:   io_writehandlers[0][port+1](port+1,(val >> 8) & 0xff,1);
src/hardware/iohandler.cpp:   io_writehandlers[1][port+0](port+0,(val >> 0 ) & 0xffff,2);
src/hardware/iohandler.cpp:   io_writehandlers[1][port+2](port+2,(val >> 16) & 0xffff,

`port` is even `& 0xffff`'d in plenty of places, which destroy values greater than 2^16 -1.  So logically (even if we ignore optimization or not), the value should be 16-bit. 

But if we don't want to do this - what's a logically approach to getting rid of wasteful Bitu use, even in obvious cases like this?

8 bytes * 6 arrays * 64k elements == over 3MB, which blows the average CPU's L3 data-cache.  Everything on the DOS side is either 16-bits (such as ports), and 32-bits (memory address), and lower.  I thought this was a sane place to start ratchetting down from the wasteful use of `Bitu`.

kcgen commented:
> In several places I noticed 16-bit port is then packaged back into Bitu for processing by other functions

Right; there will be a lot of this back-and-fourth in the mean-time while we steadily get `Bitu` stuff under control.   The only other option is a mass-update, which we've already decided we're not willing to review/make.

So how do we ratchet down, in small bite-sized chunks, without addressing comments like this - if we want all these other things fixed too (besides easy-to-review 1:1 replacements), then the PR would be massive. (This PR is ~175 LOC changes).

kcgen commented:
It might be easy just to wall off this type of work, and have you lead the charge if we're going to work on this.

dreamer commented:
> 8 bytes * 6 arrays * 64k elements == over 3MB, which alone exceeds the average CPU's L3 data-cache.

This is a very good point. Reducing size from ~3MB to ~786 KB would definitely help…

> Talking about these types of changes in general, given everything on the DOS side is either 16-bits (such as ports), and 32-bits (memory address), I thought this was a sane place to start ratchetting down from the wasteful use of `Bitu`.

The idea is very good - I would argue for doing this carefully step-by step, so we don't lock ourselves from some weird edge-case by the mistake… also, I believe we should not go straight to `uint16_t`, but rather introduce a new type, specifically for indicating port numbers, e.g.:
```
using port_t = Bitu;
```
Once this transition is done, it will be easier to change port type to the one that's the most appropriate, and having a new type definition will make it possible to do some C++ hacks to detect problematic areas. I think we could perhaps safely change to `uint32_t` (but even for that, I'm only 95% sure and would need actual numbers to put on CPU emulation speed before and after…).

Anyway - we need to learn about this part of emulator first:

- what are those +3 additional addresses going beyound 2<sup>16</sup>?
- Bochs project has long list of known I/O ports for XT, AT, and PS/2: [link](http://bochs.sourceforge.net/techspec/PORTS.LST) - it does include some memory-mapped I/O addresses like 0xC0000000 for math coprocessor - how dosbox handles those?
- what about PC98 platform? does it have the same I/O port addressing? is changing the type going to lock us out of porting the support from DOSBox-X in the future?
- what type is actually used by our CPU emulator code here? Does changing it to a different one does not mean we're actually slowing CPU emulation down (thus making the slowest part of dosbox slower to get unmeasured performance improvement through caching?)

That's a lot of work… I think we're not ready to do it yet.

kcgen commented:
@dreamer , great questions and very nice suggestion regarding the `io_port_t` type!
I haven't looked into them yet; and agree we're not ready to tackle this yet. 

Before we back-burner this, I measured _handler-calls-per-microsecond_, by calling 2,000 function pointers, of which are sitting in a 64k element array to represent IO_MAX, using realistic prototypes of either three `Bitu`s across the board vs (assumed) realistically-sized argument types (`uint16_t` port, `uint32_t` value, and `uint8_t` type length).

On 64-bit x86, the `Bitu`'s achieved ~120 calls/μs, sized-types `uint16_t` ~144 calls/μs, and the sized-and-fast `uint_fast16_t` types were roughly similar to the sized types. 

On the Pi3, I believe the array sizes playing havok with the cache, as `-Os` was always the winner.  Of these, `Bitu`'s did 43.5 calls/μs, sized-types 45.6 calls/us, and the sized-and-fast types 48.8 calls/us.

So `Bitu` didn't lose by much, but at least the results are win-win: less memory usage and a bit more speed, and perhaps played out over across even more data types we'll see cache playing a bigger roll. 

```
GCC   -Ofast: [ignore 16] Bloated types, calls per us = 120.50
GCC   -O3   : [ignore 16] Bloated types, calls per us = 120.28
GCC   -O2   : [ignore 16] Bloated types, calls per us = 114.07
GCC   -Os   : [ignore 16] Bloated types, calls per us = 117.19

GCC   -Ofast: [ignore 0] Sized types, calls per us = 144.88
GCC   -O3   : [ignore 0] Sized types, calls per us = 145.30
GCC   -O2   : [ignore 0] Sized types, calls per us = 145.42
GCC   -Os   : [ignore 0] Sized types, calls per us = 133.87

GCC   -Ofast: [ignore 48] Compiler-tuned types, calls per us = 143.98
GCC   -O3   : [ignore 48] Compiler-tuned types, calls per us = 144.62
GCC   -O2   : [ignore 48] Compiler-tuned types, calls per us = 144.94
GCC   -Os   : [ignore 48] Compiler-tuned types, calls per us = 117.39
```

[type_size_handler_benchmark.tar.gz](https://github.com/dosbox-staging/dosbox-staging/files/4699770/type_size_handler_benchmark.tar.gz)
